### ISSUE TITLE
Problem with publish function (delay)

### ISSUE THREAD

**Issue by marcoliviergagne** [knolleary/pubsubclient#798]
Hi, my program uses the Pubsubclient library to publish MQTT message at high frequency (around 110 Hz). We notice that once in a while (like 200 times per hour), the publish function takes more time than usual (usually around 2000 microseconds).  This duration can reach a few seconds (sometimes 10 seconds!) and it is problematic because I need a stable average of 110 Hz. I cannot have such delays between samples, even if these delays are occasionnal. The message sent 110 times every second is a variable 156 bytes message. 

I can identify that the delay in publish function is specifically located in the write function :

write(header,this->buffer,length-MQTT_MAX_HEADER_SIZE), 

Then more precisly in this write function this line :     rc = _client->write(buf+(MQTT_MAX_HEADER_SIZE-hlen),length+hlen).  
 
Does anybody know what could be the cause of this occasionnal delay ? The program is a firmware that is uploaded in a ESP32 sensor and sent 110 Hz MQTT message to a raspberry pi 4b.

**Comment by knolleary** [comment]
That's the call to the underlying network client to write the bytes - so the delay is coming from in there somewhere.

Without knowing what network client you are using with PubSubClient, its hard to say (although I'm not sure I'd know what to suggest even if you had said what client you are using).

Could be the network client's internal buffers are full waiting to get the bytes written to the network.... really hard to say any more than that.

**Comment by marcoliviergagne** [comment]
Ok thank you ! Is there a function to empty the buffer before using publish ? (to make sure i empty the buffer). Thank you

**Comment by knolleary** [comment]
Any buffer will be in the network client you are using. You'd have to check it's documentation for what it provides in that area.

**Comment by marcoliviergagne** [comment]
Ok thank you for your help. I am still searching for the solution (I use WiFiClient), I let you know when I find it, and do not hesitate if you have any other idea. Thanks again!

**Comment by bogus105** [comment]
marcoliviergagne - have you found solution for your delays problem?

**Comment by marcoliviergagne** [comment]
Yeah that was a hardware problem (antenna). Sorry for that. Thank you
